### PR TITLE
[WIP] Context API

### PR DESCRIPTION
**Pull request by satac2**
If you could take a look at the class declarations and see if I am on the right track for the context api, that would be great, thanks.

### CODE REVIEW

File: api/include/opentelemetry/context/context.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+{
+
+
+/*The context class provides a context identifier*/
+class Context{
+  private:
+    /*The identifier itself*/
+    std::map<std::string, int> ctx_;
+
+  public:
```
Comment (reyang) [inline api/include/opentelemetry/context/context.h:17]: Things to be considered:
1. Is this an abstract class or the actual implementation? If it is the former, need to have pure virtual functions.
2. Will we have APIs to Create a Key, Get Value and Set Value https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/context.md#create-a-key.

File: api/include/opentelemetry/context/context.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    /* Context: contructor, creates a context object from a map
+     * of keys and identifiers
+     */
+    Context(std::map<std::string, int> ctx);
```
Comment (reyang) [inline api/include/opentelemetry/context/context.h:21]: All the public APIs need to be ABI compatible - which means they are not supposed to expose dependency on STL traits/layout.

File: api/include/opentelemetry/context/context.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    /* RuntimeContext: A constructor that will set the current
+     * context to the threading local.
+     */
+    RuntimeContext();
+
+    /* attach: Sets the current 'Context' object. Returns a token 
+     * that can be used to reset to the previous Context.
+     */
+    static Token Attach(Context context);
```
Comment (reyang) [inline api/include/opentelemetry/context/context.h:46]: Do we want the context to be a plain copy, a shallow copy, reference or a shared pointer?

File: api/include/opentelemetry/context/context.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+
+/*The context class provides a context identifier*/
+class Context{
+  private:
+    /*The identifier itself*/
+    std::map<std::string, int> ctx_;
```
Comment (reyang) [inline api/include/opentelemetry/context/context.h:15]: Need to put this into consideration:
* According to the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/context.md#overview), "A Context MUST be immutable, and its write operations MUST result in the creation of a new Context containing the original values and the specified values updated".

**Comment by reyang** [comment]
Welcome to your first PR @satac2, please clear the CLA steps as required by the CI / CNCF.

File: api/include/opentelemetry/context/context.h [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        
+        context_ = token.GetContext();
+         
+        return 0;     
+      }
+
+
+    private:
+     
+      static thread_local Context context_;
```
Comment (maxgolov) [inline api/include/opentelemetry/context/context.h:224]: Not 100% sure this static would always work in certain DLL loading scenarios, where two different DLLs consume OT API surface, esp. lazy loading / delay-loading. May want to spell out the limitations somewhere in the class documentation..

https://devblogs.microsoft.com/oldnewthing/20101122-00/?p=12233

Comment (pyohannes) [reply api/include/opentelemetry/context/context.h:224]: Agreed. Rather use this [singleton style](http://laristra.github.io/flecsi/src/developer-guide/patterns/meyers_singleton.html) (adding `thread_local`). That's what we use [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/provider.h#L46).

File: api/include/opentelemetry/context/context.h [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+  std::mutex context_id_mutex;
+
+  /*The context class provides a context identifier */
```
Comment (g-easy) [inline api/include/opentelemetry/context/context.h:15]: Please use // style comments and add a full stop. :)

File: api/include/opentelemetry/context/context.h [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+            key_name_ = key_name;
+            identifier_ = identifier;
+          }
+
+        public:
+          
+          /* Consructs a new ContextKey with the passed in name and increments
+           * the identifier then assigns it to be the key's identifier.
+           */
+          ContextKey(std::string key_name){
```
Comment (g-easy) [inline api/include/opentelemetry/context/context.h:51]: Please mark this constructor explicit.

File: api/test/context/BUILD [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    deps = [
+        "//api",
+        "@com_google_googletest//:gtest_main",
+    ],
+)
+
+cc_test(
+    name = "runtimeContext_test",
+    srcs = [
+        "runtimeContext_test.cc",
```
Comment (g-easy) [inline api/test/context/BUILD:17]: Please use lowercase + underscores, i.e. runtime_context_test (also on L15) to be consistent with the rest of the files.

File: api/test/context/BUILD [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    ],
+    copts = ["-Wall","-std=c++17"],
```
Comment (g-easy) [inline api/test/context/BUILD:19]: Please remove these copts.

**Comment by g-easy** [comment]
More general context questions:

1. Should the actual TLS part be a pointer, and allocate the per-thread context when it's accessed by Attach/GetCurrent?

2. However we end up resolving the std::map question: how can we attach/detach faster than doing a full copy of the map?

**Comment by reyang** [comment]
> 2. However we end up resolving the std::map question: how can we attach/detach faster than doing a full copy of the map?

https://www.python.org/dev/peps/pep-0550/#implementation 😄

**Comment by g-easy** [comment]
And refcount the context objects that form chains?

**Comment by g-easy** [comment]
Can we have a no-op Context API and move the implementation into the SDK, so that other SDKs can make different performance tradeoffs in how this is implemented?

**Comment by reyang** [comment]
> Can we have a no-op Context API and move the implementation into the SDK, so that other SDKs can make different performance trade-offs in how this is implemented?

Probably not, the OpenTelemetry API package has to give a default implementation that flows the incoming traceid/spanid/tracestate/flags to the outgoing request, thus the actual implementation cannot be in the SDK.
However we need to make sure that other SDKs / components can replace the context implementation.

File: api/include/opentelemetry/context/context.h [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+      /* A default constructor that will set the context to
+       * an empty context object.
+       */
+      RuntimeContext(){
+        context_ = Context();
+      }
+
+      /* A constructor that will set the context as the passed in context. */
+      RuntimeContext(Context &context){
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/context/context.h:182]: I'm a little confused what you mean by this, should the ThreadLocalContext be a derived class of RuntimeContext?

[the remark this answers is not in the document]
Comment (reyang) [reply api/include/opentelemetry/context/context.h:182]: > I'm a little confused what you mean by this, should the ThreadLocalContext be a derived class of RuntimeContext?

Yes.

[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/context/context.h:182]: How would it be different than RuntimeContext, should the RuntimeContext just have a static context and the ThreadLocal have a static thread_local context?

[the remark this answers is not in the document]
Comment (reyang) [reply api/include/opentelemetry/context/context.h:182]: I think RuntimeContext is the abstraction layer (which doesn't have understanding about TLS).

**Pull request closed without merging**